common/src/jni/main/cpp/conscrypt/native_crypto.cc

```diff
@@ -7048,6 +7048,97 @@ static jbyteArray NativeCrypto_SSL_get_tls_unique(JNIEnv* env, jclass, jlong ssl
+        for (int i = 0; i < paramsValues.size(); i++) {
+            paramsBytes.get()[i] = static_cast<uint8_t>(paramsValues[i]);
+        }
+        ret = SSL_set_token_binding_params(ssl, paramsBytes.get(), paramsValues.size());
+    }
+
+    JNI_TRACE("ssl=%p NativeCrypto_SSL_set_token_binding_params => %d", ssl, ret);
+
+    if (!ret) {
+        conscrypt::jniutil::throwSSLExceptionStr(env, "Could not set token binding parameters");
+        ERR_clear_error();
+    }
+}
+
+static int NativeCrypto_SSL_get_token_binding_params(JNIEnv* env, jclass, jlong ssl_address,
+        CONSCRYPT_UNUSED jobject ssl_holder) {
+    CHECK_ERROR_QUEUE_ON_RETURN;
+    SSL* ssl = to_SSL(env, ssl_address, true);
+    JNI_TRACE("ssl=%p NativeCrypto_SSL_get_token_binding_params", ssl);
+    if (ssl == nullptr) {
+        return 0;
+    }
+    int ret;
+    if (!SSL_is_token_binding_negotiated(ssl)) {
+        ret = -1;
+    } else {
+        ret = SSL_get_negotiated_token_binding_param(ssl);
+    }
+    JNI_TRACE("ssl=%p NativeCrypto_SSL_set_token_binding_params => %d", ssl, ret);
+    return ret;
+}
+
+static jbyteArray NativeCrypto_SSL_export_keying_material(JNIEnv* env, jclass, jlong ssl_address,
+        CONSCRYPT_UNUSED jobject ssl_holder, jbyteArray label, jbyteArray context, jint num_bytes) {
+    CHECK_ERROR_QUEUE_ON_RETURN;
+    SSL* ssl = to_SSL(env, ssl_address, true);
+    JNI_TRACE("ssl=%p NativeCrypto_SSL_export_keying_material", ssl);
+    if (ssl == nullptr) {
+        return nullptr;
+    }
+    ScopedByteArrayRO labelBytes(env, label);
+    if (labelBytes.get() == nullptr) {
+        JNI_TRACE("ssl=%p NativeCrypto_SSL_export_keying_material label == null => exception", ssl);
+        return nullptr;
+    }
+    std::unique_ptr<uint8_t[]> out(new uint8_t[num_bytes]);
+    int ret;
+    if (context == nullptr) {
+        ret = SSL_export_keying_material(ssl, out.get(), num_bytes,
+                        reinterpret_cast<const char*>(labelBytes.get()), labelBytes.size(),
+                        nullptr, 0, 0);
+    } else {
+        ScopedByteArrayRO contextBytes(env, context);
+        if (contextBytes.get() == nullptr) {
+            JNI_TRACE("ssl=%p NativeCrypto_SSL_export_keying_material context == null => exception", ssl);
+            return nullptr;
+        }
+        ret = SSL_export_keying_material(ssl, out.get(), num_bytes,
+                        reinterpret_cast<const char*>(labelBytes.get()), labelBytes.size(),
+                        reinterpret_cast<const uint8_t*>(contextBytes.get()), contextBytes.size(), 1);
+    }
+    if (!ret) {
+        conscrypt::jniutil::throwExceptionFromBoringSSLError(env, "SSL_export_keying_material",
+                conscrypt::jniutil::throwSSLExceptionStr);
+        JNI_TRACE("ssl=%p NativeCrypto_SSL_export_keying_material => exception", ssl);
+        return nullptr;
+    }
+    jbyteArray result = env->NewByteArray(static_cast<jsize>(num_bytes));
```

Comment:
If this allocation returns `null`, it would seem that we should throw something, no? Maybe not so severe as an `OOM`, but it seems exceptional.  Not sure how consistent we are with this case, but it might be worth revisiting how we handle this throughout.

Reply:
Yeah, that's reasonable.  I went with SSLException since that's already declared to be thrown.  (We're definitely inconsistent in what we do in response to NewFooArray failures.)